### PR TITLE
Support writeAtomicBlob from InputStream for repository blob container interface

### PR DESCRIPTION
**Pull request by kingherc**
Add writeAtomicBlob from InputStream to the container interface, mostly for fs and hdfs repos (since cloud repos write atomically), similar to how writeAtomicBlob from bytes is implemented for fs and hdfs already (write temp file and rename atomically).

Relates ES-9248

### CODE REVIEW

File: test/framework/src/main/java/org/elasticsearch/snapshots/mockstore/MockRepository.java
Comment (ywangd) [inline test/framework/src/main/java/org/elasticsearch/snapshots/mockstore/MockRepository.java:671]: Nit: It seems to me that we can get rid of the other variant now since the default implementation from `BlobContainer` delegates it to this one.

File: x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/analyze/BlobAnalyzeAction.java
Comment (ywangd) [inline x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/analyze/BlobAnalyzeAction.java:347]: Nit: I am not sure whether this randomization is really necessary. This is not a test and we don't randomize for `writeBlob` a few lines below?

**Review by ywangd — APPROVED** [review]
LGTM

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java
Comment (arteam) [inline modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java:302]: Is it intentional to remove this assertion? It seems to me that regardless of which `writeBlobAtomic` method we use, `OperationPurpose` and blob name stay the same and can be checked for consistency?

**Review by arteam — APPROVED** [review]
LGTM with one question

File: test/framework/src/main/java/org/elasticsearch/snapshots/mockstore/MockRepository.java
Comment (kingherc) [reply test/framework/src/main/java/org/elasticsearch/snapshots/mockstore/MockRepository.java]: We cannot remove the other `writeBlobAtomic` a few lines below, because it also has logic to simulate failures. But we can make the other BytesReference-based variant call this InputStream-based variant. Did that.

File: x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/analyze/BlobAnalyzeAction.java
Comment (kingherc) [reply x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/analyze/BlobAnalyzeAction.java]: Ah yes! Will revert.

File: modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java
Comment (kingherc) [reply modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java:302]: Yes it's intentionally removed because I found it superfluous because the `writeBlob` function it calls already does `assert assertPurposeConsistency(purpose, blobName);` in its beginning.

**Review by kingherc — COMMENTED** [review]
Thanks for the reviews! Will merge once the serverless PR is also reviewed.

**Comment by kingherc** [comment]
CI failure is https://github.com/elastic/elasticsearch/issues/112374 . Will retry.

File: server/src/main/java/org/elasticsearch/common/blobstore/BlobContainer.java
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/common/blobstore/BlobContainer.java:155]: Can we strengthen this to say that it must be the amount of bytes in the input stream? Not asking to verify that (we do for S3 multi-part at least though), but this is hardly a specification of the parameter.

File: server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java:317]: I think we could easily make `writeToPath` return the number of bytes written and then verify that it matches `blobSize`. That would be nice to verify that `blobSize` is filled in correctly.

**Review by henningandersen — COMMENTED** [review]
Looks good, couple of comments to address though.

File: server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java
Comment (kingherc) [reply server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java:317]: I guess this should be an assertion, right? Added one in writeToPath() itself, and also a similar assertion in the HDFS blob container.

**Review by kingherc — COMMENTED** [review]
Thanks @henningandersen ! Done.

File: server/src/main/java/org/elasticsearch/common/blobstore/BlobContainer.java
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/common/blobstore/BlobContainer.java]: I am not sure the new contract will hold for the existing spot (at least I did not check), perhaps we should avoid changing that for now (unless you did a thorough check of it?).

File: server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java:317]: Assertion is great, thanks. Not sure if we can apply it in `writePath` though, depends on my other comment.

File: server/src/main/java/org/elasticsearch/common/blobstore/BlobContainer.java
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/common/blobstore/BlobContainer.java]: We can also see what CI thinks ofc.

File: server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java
Comment (kingherc) [reply server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java:317]: Which `writePath` you mean?

File: server/src/main/java/org/elasticsearch/common/blobstore/BlobContainer.java
Comment (kingherc) [reply server/src/main/java/org/elasticsearch/common/blobstore/BlobContainer.java]: It should hold for that, as writeBlob (non-atomic) also writes the same amount of bytes, just in a non-atomic fashion (e.g., another thread might read partial results). Lets see what CI thinks.

**Review by kingherc — COMMENTED** [review]
Thanks @henningandersen . Please note my intention is to merge this and the serverless PR once they pass through CI. Feel free to tell me whether I should wait for your explicit approvals in both PRs before merging them.

File: server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java:317]: Sorry, meant `writeToPath`. Since that is used both for `writeBlob` and `writeBlobAtomic`, we may not be able to assert blob-size is right there, but this all depends on the examination in the other comment.

File: server/src/main/java/org/elasticsearch/common/blobstore/BlobContainer.java
Comment (kingherc) [reply server/src/main/java/org/elasticsearch/common/blobstore/BlobContainer.java]: I manually checked our Azure, GCP, S3, FS implementations and they use the right blob-size rather than a different one. So I think the new comment and check hold.

**Review by henningandersen — APPROVED** [review]
LGTM.

**Pull request merged**
